taqueria-vscode-extension-web-ui/src/data-editors/MichelineEditor.tsx

```diff
@@ -0,0 +1,228 @@
+	originalFormat: 'json';
+	value: object;
+};
+
+const json2Micheline = (v: object) => {
+	let michelson: string | undefined = undefined;
+	const expression = parser.parseJSON(v);
+	michelson = emitMicheline(expression);
+	return michelson;
+};
+
+const micheline2Json = (v: string) => {
+	let json: object | undefined | null = undefined;
+	json = parser.parseMichelineExpression(v);
+	return json;
+};
+
+export const getMicheline = (value: MichelineValue | undefined) => {
+	if (!value) {
+		return undefined;
+	}
+	if (value.originalFormat === 'micheline') {
+		return value.value;
+	}
+	return json2Micheline(value.value);
+};
+
+export const getJson = (value: MichelineValue | undefined) => {
+	if (!value) {
+		return undefined;
+	}
+	if (value.originalFormat === 'json') {
+		return value.value;
+	}
+	return micheline2Json(value.value);
+};
+
+export const getFriendlyDataType = (dataType: any): string => {
+	if (!isObject(dataType) || !dataType.prim) {
+		return `Error: ${JSON.stringify(dataType)}`;
+	}
+	let friendlyDataType: string;
+	switch (dataType.prim) {
+		case 'list':
+		case 'set':
+		case 'option':
+			friendlyDataType = `${dataType.prim} of ${getFriendlyDataType(dataType.args[0])}`;
+			break;
+		case 'unit':
+			return 'unit';
+		case 'string':
+		case 'nat':
+		case 'int':
+		case 'bytes':
+		case 'timestamp':
+		case 'mutez':
+		case 'address':
+		case 'key':
+		case 'key_hash':
+		case 'signature':
+		case 'chain_id':
+		case 'bls12_381_g1':
+		case 'bls12_381_g2':
+		case 'bls12_381_fr':
+		case 'sapling_transaction ms':
+		case 'sapling_state ms':
+		case 'ticket':
+		case 'chest':
+		case 'chest_key':
+		case 'tx_rollup_l2_address':
+		case 'bool':
+			friendlyDataType = dataType.prim;
+			break;
+		case 'pair':
+			friendlyDataType = `${dataType.prim} of ${(dataType.args as any[]).map(x => getFriendlyDataType(x)).join(', ')}`;
+			break;
+		case 'map':
+		case 'big_map':
+			friendlyDataType = `${dataType.prim} from ${getFriendlyDataType(dataType.args[0])} to ${
+				getFriendlyDataType(dataType.args[1])
+			}`;
+			break;
+		default:
+			return `Error: ${JSON.stringify(dataType)}`;
+	}
+	if (dataType.annots && dataType.annots.length) {
+		return `${(dataType.annots as string[]).map(x => x?.substring(1)).join(' ')} (${friendlyDataType})`;
+	}
+	return friendlyDataType;
+};
+
+declare const vscodeApi: WebviewApi<any>;
+
+export const MichelineEditor = (
+	{ input: { dataType, value, actionTitle, showDiagnostics }, onMessage }: {
+		input: { dataType: any; value?: MichelineValue; actionTitle?: string; showDiagnostics?: boolean };
+		onMessage: MichelineEditorMessageHandler;
+	},
+) => {
+	const previousState = vscodeApi.getState() as MichelineValue | undefined;
+
+	const [currentState, setState] = useState({
+		value: previousState ?? value,
+		showDiagnostics: showDiagnostics ?? false,
+	});
+
+	const handleChange = (v: MichelineValue) => {
+		setState({
+			value: v,
+			showDiagnostics: currentState.showDiagnostics,
+		});
```

Comment:
Use function set state syntax

Reply:
Can you provide more explanation?

Reply:
```ts
setState(prevState => ({
			value: v,
			showDiagnostics: prevState .showDiagnostics,
}));
```

Reply:
Why is this better?

Reply:
You can have race conditions as react might batch multiple `setState` calls together so they all read old outdated value:
https://reactjs.org/docs/hooks-reference.html#functional-updates